runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java

```diff
@@ -1296,6 +1302,30 @@ private StreamingPubsubIOWriteOverrideFactory(DataflowRunner runner) {
   }
 
   @VisibleForTesting
+  static class StreamingShardedWriteFactory<T>
+      implements PTransformOverrideFactory<PCollection<T>, PDone, WriteFiles<T>> {
+    DataflowPipelineWorkerPoolOptions options;
+
+    StreamingShardedWriteFactory(PipelineOptions options) {
+      this.options = options.as(DataflowPipelineWorkerPoolOptions.class);
+    }
+
+    @Override
+    public PTransformReplacement<PCollection<T>, PDone> getReplacementTransform(
+        AppliedPTransform<PCollection<T>, PDone, WriteFiles<T>> transform) {
+        return PTransformReplacement.of(
+            PTransformReplacements.getSingletonMainInput(transform),
+            transform.getTransform().withNumShards(options.getMaxNumWorkers() * 2));
```

Comment:
Add a comment here explaining why this is a good choice, and what effect it will have? (how many output files we'll actually get, and how many we would have gotten without this override)